### ISSUE TITLE
Add modeling rule for joining too many models

### ISSUE THREAD

**Issue by cmcnicoll** [dbt-labs/dbt-project-evaluator#394]
### Describe the feature
Per the [best practices docs](https://docs.getdbt.com/best-practices/how-we-structure/3-intermediate#intermediate-models):

> Structural simplification. Bringing together a reasonable number (typically 4 to 6) of entities or concepts (staging models, or perhaps other intermediate models) that will be joined with another similarly purposed intermediate model to generate a mart — rather than have 10 joins in our mart, we can join two intermediate models that each house a piece of the complexity, giving us increased readability, flexibility, testing surface area, and insight into our components.

**Comment by b-per** [comment]
Would you be keen to take this one on?

**Comment by cmcnicoll** [comment]
Thanks for the response @b-per. Sorry, I don't have time right now.

**Comment by dave-connors-3** [comment]
Hey brad! yep that's about it! it would look like:

1. adding a `fct_` model to the `modeling` marts directory, with logic like you described! I might suggest lowering the barrier to 7 or more parent models :) 
2. Adding the `is_empty` test to that model (as other models have) to raise the expected warning
3. adding a model in the `integration_tests` folder that violates this rule, to ensure it's working as intended. 
4. Add a seed to the `integration_tests/seeds` directory that matches the expected output of the `fct_` model you added


Let me know if that makes sense! happy to help shepherd that along

**Comment by BradCr** [comment]
Thanks, Dave.

Where you mention adding the new `fct_` model to the `modeling` marts directory, I don't see a "modeling" directory under `models/marts/`. Is the `/models/marts/dag` directory a good place for the new fct_ model?

I will also need to add information about this model to the `docs/rules/modeling.md` file, right?

**Comment by BradCr** [comment]
I've completed making the changes as you've mentioned. In addition, I updated the docs to explain the new model. For the screenshots I had two concerns; When I attached the screenshots they are linked to my repository (not sure if that gets automatically updated or if it's something that a dbt maintainer would have to deal with) and the screenshots are from the new Cloud Lineage diagram, so the look and feel is different from the existing screenshots.

The thing I haven't been able to figure out is how to run the integration_tests. I am working on VSCode and cloned my repository, I am able to build and run the new model (on Duckdb) but I can't figure out how to have it run the models in the integration_tests. Are there instructions on how to do that? The readme for this package is kind of light on details of how to actually do that test.

Otherwise, I think it's ready for review. Let me know if I should create the PR now, or if I need to figure out how to run the integration_tests and do that first.

Thanks!

**Comment by BradCr** [comment]
@dave-connors-3 Sorry for the direct ping, I'm not sure if you get notified when a comment gets posted or only if you're tagged.

When you have time, can you look at my last 2 comments?

Thanks.

**Comment by dave-connors-3** [comment]
hey @BradCr! sorry for the delay!

> Where you mention adding the new fct_ model to the modeling marts directory, I don't see a "modeling" directory under models/marts/. Is the /models/marts/dag directory a good place for the new fct_ model?

yes, dag! was going off of memory which is never a good idea

second off, would love if you opened a draft PR with your changes! Will definitely help to see your changes to speak a bit more intelligently! 

thirdly, integration tests! if whatever i come up with here works for you, i'll add it to the docs for future contributors :) 

The basic idea is that the `integration_tests` folder contains a dbt project with a bunch of dummy models that violate a whole bunch of the rules. This project imports the actual `dbt_project_evaluator` project as a local package, so when you run it, you'll build all the models in `integration_tests` as well as the ones from project evaluator. Then, there's a seed in the `seeds` folder for each of `fct` models that asserts what the output of the model should be from the `integration_tests` project. For example [`test_fct_missing_primary_key_tests`](https://github.com/dbt-labs/dbt-project-evaluator/blob/main/integration_tests/seeds/tests/test_fct_missing_primary_key_tests.csv) asserts what the output of [`fct_missing_primary_key_tests`](https://github.com/dbt-labs/dbt-project-evaluator/blob/main/models/marts/tests/fct_missing_primary_key_tests.sql) should be given the models in the `integration_tests` project. This seed is configured with a [dbt test](https://github.com/dbt-labs/dbt-project-evaluator/blob/main/integration_tests/seeds/tests/tests_seeds.yml#L6) to make that assertion! When all this is configured, a passing `dbt build` on this project means your integration tests are passing!

to run the integration tests: 
1. if you haven't already, install `dbt-duckdb` (seems like you're good here) 
2. add an `integration_tests` profile to your profiles.yml
```yml
integration_tests:
  outputs:
    duckdb:
      type: duckdb
      path: ./dbt.duckdb
  target: duckdb
```
3. add/edit the models in the project to introduce a violation that would be caught by your new rule
4. add a seed called `test_{{ your new fct model name }}` in the appropriate directory that contains the expected output of your new fct model
5. configure the equality test like in the above example
6. `dbt build` 

A note that adding + editing models may cause failures in other integration tests in the case that you accidentally introduce violations of other rules. You can edit as necessary to get the output you expect. 

that all make sense?

**Comment by BradCr** [comment]
Thanks, @dave-connors-3, it's all making sense to me. 

I'll give the integration tests another try and let you know how it goes. I think the only thing I hadn't yet done was the equality test between the seed (which I did create) and the model that violated the new rule (which I also did create). I didn't modify any of the existing seeds or violating-models.

I'll also open up the draft PR.

Thanks again for your help, I'm excited to keep moving.

**Comment by BradCr** [comment]
So, here's a very bizarre thing I found out. In the root of the dbt-project-explorer I could run dbt deps just fine, but when I went into the integration_tests subfolder and ran dbt deps there, it would just hang at "Installing". That's why I wasn't able to do a dbt build inside that folder. I finally looked at the dbt.log and found this error message: "Symlinks are not available on this OS, copying dependency."

A Google search led me to this old Git issue: https://github.com/dbt-labs/dbt-core/issues/766, which talks about Admin user level access being required to run Symlinks on Windows. So when I ran VS Code as an Admin, the dbt deps ran just fine. Weird!

Anyhow, I'm now working on running the test and then I'll open the PR.
